### PR TITLE
feat(sql): update  "within" operator to accept GeoHash literals and constant functions

### PR DESCRIPTION
**Pull request by eugenels**

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 1]
```diff
@@ -1158,6 +1172,80 @@ private boolean removeWithin(AliasTranslator translator, ExpressionNode node, Re
...
+            final boolean isCharsPrefix = len > 1 && token.charAt(0) == '#';
+
+            if (!(isConstant && (isBitsPrefix || isCharsPrefix))) {
+                throw SqlException.$(position, "GeoHash literal expected");
+            }
+
+            final int type;
+            final long hash;
+            try {
+                if (token.charAt(1) != '#') {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1214]: minor: can be `if (isCharsPrefix)`

**Review by ideoma — APPROVED** [review]
Looks good. We will probably need another pass to accept Bind Variables

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 85 / 88 (96.59%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|59|62|95.16%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 85 / 88 (96.59%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|59|62|95.16%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java (hunk 1 above)
Comment (eugenels) [reply core/src/main/java/io/questdb/griffin/WhereClauseParser.java]: updated

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 83 (96.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|54|57|94.74%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 83 (96.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|54|57|94.74%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 2]
```diff
@@ -1158,6 +1172,78 @@ private boolean removeWithin(AliasTranslator translator, ExpressionNode node, Re
...
+        final int type;
+        final long hash;
+
+        if (isFunc(inArg)) {
+            Function f = functionParser.parseFunction(inArg, metadata, executionContext);
+            if (isGeoHashConstFunction(f)) {
+                type = f.getType();
+                hash = GeoHashes.getGeoLong(type, f, null);
+            } else {
+                throw SqlException.$(inArg.position, "GeoHash const function expected");
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1198]: function must be closed otherwise it could cause memory leak

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 82 (96.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|53|56|94.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 82 (96.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|53|56|94.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 82 (96.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|53|56|94.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Review by marregui — COMMENTED** [review]
`select * from pos latest by uuid where time in '2021-05-11' and hash within ('z', 'b')`

hehe, no longer possible to drop a string instead of a geohash. In other places we are more lenient, for instance in eq comparison we cast the str to a geohash. Would it make sense to allow this here? I do not have strong opinions in here.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 82 (96.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|53|56|94.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 82 (96.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|53|56|94.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 82 (96.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|53|56|94.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 79 / 82 (96.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|53|56|94.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Pull request merged**